### PR TITLE
PEX (Port Expander) library

### PR DESCRIPTION
**Pull request by brunofalmeida**

### CODE REVIEW

**Comment by brunofalmeida** [comment]
@siddMahen I ran the test and it's all working. Ready to merge besides any cleanup.

**Comment by siddMahen** [comment]
I'm going to re-write this using the independent object interface idea.

**Comment by brunofalmeida** [comment]
Sounds good, just keep in mind that unlike the ADC, this is not something where data needs to be transmitted over CAN to the OBC.

**Comment by siddMahen** [comment]
What is `pex_addr` used for? Is it always the same for the lifetime of the pex?

**Comment by siddMahen** [comment]
It seems like `pex_addr` is just constant for all pex ops on, say, the payload board.

**Comment by brunofalmeida** [comment]
Each PEX device has a 3-bit address that's set by how you wire the 3 hardware pins. The address is given as part of the SPI message data, which allows you to have up to 8 PEX devices connected to the same SPI pins including CS.

**Comment by siddMahen** [comment]
Ah I see; thanks. @brunofalmeida Also, why are there delays when reading and writing?

**Comment by brunofalmeida** [comment]
Not sure, might have to do with minimum timing delays? @vogeldylan would probably know

**Comment by siddMahen** [comment]
10 ms seems pretty long for a timing delay; were those required to make it work reliably, or were they left there by the last guy testing? @vogeldylan

**Comment by siddMahen** [comment]
Also, @vogeldylan @brunofalmeida wasn't there some issue with pex whereby we could avoid resetting frequently by keeping track of what we're doing; wasn't this related to the fancy ADC in some way? What exactly was the issue? Has it been fixed?

**Comment by brunofalmeida** [comment]
That issue was that the optical ADC (more powerful/fancy one for optical sensors) had its CS pin connected to a PEX, which would send unintended commands to the ADC that were intended for the PEX. We had to reset the ADC every time we used it. That problem is fixed in the new layout of the PAY-Optical PCB.

**Comment by siddMahen** [comment]
@brunofalmeida Rewrote most of the PEX lib; still need to

* Add documentation for `pex_get_pin()` which I forgot to port
* Re-write and re-test the func_test code using the new interface

File: include/pex/pex.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#ifndef PEX_H
+#define PEX_H
+
+#include <spi/spi.h>
+#include <uart/uart.h>
```
Comment (siddMahen) [inline include/pex/pex.h]: UART dependence should be removed

**Comment by siddMahen** [comment]
@brunofalmeida The new interface is slightly less intuitive (you need to know the function argument order), but it saves on the size of the library and the amount of code that needs to compiled in. Also makes it much easier to catch bugs and make improvements.

**Comment by vogeldylan** [comment]
Yeah I thought the 10ms delays were removed in one of my earlier edits to the library, not sure how those showed up again. 

Code looks like it will work, seems a bit more straightforward than the old version. My only nit-picky complaint is that the distinction between GPIO bank A & B is called 'side', which seems un-intuitive. These banks don't have much relation to sides of the device.

**Comment by siddMahen** [comment]
@vogeldylan Ok, will change by tonight. I wasn't sure exactly what to call it; I'll use `adc_bank_t` instead.

**Comment by brunofalmeida** [comment]
@siddMahen should be `pex_bank_t`

**Comment by brunofalmeida** [comment]
@siddMahen done my changes

**Comment by siddMahen** [comment]
@brunofalmeida Just waiting for a final test, then this can be merged.

**Comment by siddMahen** [comment]
@brunofalmeida Rebased ontop of master and updated the commit message. Can you check out the commit message and make sure everything is correct?

**Comment by brunofalmeida** [comment]
Each bank contains 8 pins, not 12. Other than that it looks good, just need to run the test again.

**Comment by brunofalmeida** [comment]
@siddMahen Tested and working. Ready to merge?

**Pull request merged**